sycl/source/detail/online_compiler/online_compiler.cpp

```diff
@@ -0,0 +1,208 @@
+//==----------- online_compiler.cpp ----------------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/INTEL/online_compiler.hpp>
+#include <CL/sycl/detail/defines_elementary.hpp> // for __SYCL_INLINE_NAMESPACE
+#include <CL/sycl/detail/os_util.hpp>
+#include <CL/sycl/detail/pi.hpp>
+
+#include <cstring>
+
+#include "ocloc_api.h"
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace INTEL {
+namespace detail {
+
+static std::vector<const char *>
+prepareOclocArgs(sycl::info::device_type DeviceType, device_arch DeviceArch,
+                 bool Is64Bit, const std::string &DeviceStepping,
+                 const std::string &UserArgs) {
+  std::vector<const char *> Args = {"ocloc", "-q", "-spv_only", "-device"};
+
+  if (DeviceType == sycl::info::device_type::gpu) {
+    switch (DeviceArch) {
+    case device_arch::gpu_gen9_5:
+      Args.push_back("cfl");
+      break;
+
+    case device_arch::gpu_gen11:
+      Args.push_back("icllp");
+      break;
+
+    default:
+      Args.push_back("skl");
+    }
+  } else {
+    Args.push_back("skl");
+  }
+
+  if (DeviceStepping != "") {
+    Args.push_back("-revision_id");
+    Args.push_back(DeviceStepping.c_str());
+  }
+
+  Args.push_back(Is64Bit ? "-64" : "-32");
+
+  if (UserArgs != "") {
+    Args.push_back("-options");
+    Args.push_back(UserArgs.c_str());
+  }
+
+  return Args;
+}
+
+static std::vector<byte>
+compileToSPIRV(const std::string &Source, sycl::info::device_type DeviceType,
+               device_arch DeviceArch, bool Is64Bit,
+               const std::string &DeviceStepping, void *&CompileToSPIRVHandle,
+               void *&FreeSPIRVOutputsHandle,
+               const std::vector<std::string> &UserArgs) {
+
+  if (!CompileToSPIRVHandle) {
+#ifdef __SYCL_RT_OS_WINDOWS
+    std::string OclocLibraryName = "ocloc64.dll";
```

Comment:
I don't know what you mean by "magic" ), but I will add the 'static const' before 'std::string'

Reply:
I meant e.g. this: https://blog.mbedded.ninja/programming/languages/c-plus-plus/magic-statics/

Reply:
The _reference_ to std::string is not required in this case. Thus having the variables as static const std::string is sufficient for the purpose of using this string constant. Do you agree?

Reply:
Not sure how reference to string is relevant here. My point is every time the function is invoked, an string constructor will be executed. Magic static is usual way to avoid this. Global-scope std::string constant should be OK too here.

Reply:
static std::string object is initialized only once and is similar to global scope std::string.
12:33:37 > cat str_test.cpp 
#include < iostream >
#include < string >
struct my_string : public std::string {
  my_string(const char *Arg) : std::string(Arg) {
    std::cout << "my_string constructor is called\n";
  };
  const std::string& as_string() const { return *this; }
};
void func(int N) {
  static const my_string MyString("hello");
  std::cout << MyString.as_string() + std::to_string(N) + "\n";
}
int main() {
  for (int I = 0; I < 10; I++)
    func(I);
}
12:33:56 > clang++ str_test.cpp -O0 -o str_test.exe
12:33:59 > ./str_test.exe 
my_string constructor is called
hello0
hello1
hello2
hello3
hello4
hello5
hello6
hello7
hello8
hello9

Reply:
Right, function-local static variable is the key element of the magic static pattern. But what I meant is:
```
const std::string& getOclocLibName() {
#ifdef __SYCL_RT_OS_WINDOWS
    static std::string OclocLibraryName = "ocloc64.dll";
#else
    static std::string OclocLibraryName = "libocloc64.so";
#endif
    return OclocLibraryName;
}
```